### PR TITLE
Recovery: Allow for file recovery policies based on scope #5928

### PR DESCRIPTION
**Pull request by ChristophAmes**
Currently, file recovery policies can be set for individual file types. This will be expanded to also allow for policies to be set for an entire scope. Policies based on file types take priority over policies based on scopes.

This change will be applied to the suspicious replica recoverer.

### CODE REVIEW

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 1]
```diff
@@ -131,10 +126,11 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
         json_data = json.load(json_file)
     except ValueError:
         logger(logging.WARNING, "No JSON object could be decoded.")
+        return True
 
     # Checking that the json file is formatedd properly.
     for i, entry in enumerate(json_data):
-        if "datatype" not in entry or "action" not in entry:
+        if (("datatype" or "scope") not in entry or "action" not in entry):
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Are you sure about the test ?
Shouldn't it be : 
```python
if ("datatype" not in entry or "scope" not in entry or "action" not in entry):
```

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 2]
```diff
@@ -196,8 +192,8 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
                 logger(logging.DEBUG, '%s', i)
 
             # RSEs that aren't available shouldn't have suspicious replicas showing up. Skip to next RSE.
-            if not rse['availability_read'] and ((len(suspicious_replicas_avail_elsewhere) > 0) or (len(suspicious_replicas_last_copy) > 0)):
-                logger(logging.WARNING, "%s is not available (availability_read: %s), yet it has suspicious replicas. Please investigate. \n", rse_expr, rse['availability_read'])
+            if (rse['availability'] not in {4, 5, 6, 7}) and ((len(suspicious_replicas_avail_elsewhere) > 0) or (len(suspicious_replicas_last_copy) > 0)):
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: You should keep the code using `rse['availability_read']` that now replace the old one based on `rse['availability']` + mask

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 3]
```diff
@@ -338,49 +334,101 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
...
                         del recoverable_replicas[vo][rse_key][replica_key]
                     else:
                         # Deal with replicas based on their metadata.
                         if file_metadata["datatype"] is None:  # "None" type has no function "split()"
                             files_to_be_ignored.append(recoverable_replicas[vo][rse_key][replica_key])
+                            logger(logging.WARNING, "RSE: %s, replica name %s, surl %s: Replica does not have a data type associated with it. No action will be taken.",
+                                   rse_key, replica_key, recoverable_replicas[vo][rse_key][replica_key]['surl'])
                             continue
+
+                        file_metadata_datatype = file_metadata["datatype"].split("_")[-1]
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: In case of `DAOD_PHYS` it would select `PHYS` which is probably not what you wanted to do. Morever `_` in datatype name is very ATLAS specific. I would just keep `datatype` as it is, but rely on regular expression

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 4]
```diff
@@ -338,49 +334,101 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
...
+                        file_metadata_scope = str(file_metadata["scope"])
+                        in_file_metadata_datatype = False
+                        in_file_scope = False
+                        in_scope_wildcard = False
+                        action = ""
                         for i in json_data:
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py:365]: i is already used as a variable in line 132. In addition shouldn't it be `entry` instead of `json_data`

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 5]
```diff
@@ -338,49 +334,101 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
...
-                                            break
-                                else:
-                                    logger(logging.WARNING, "RSE: %s, replica name %s, surl %s: Match for the metadata 'datatype' (%s) of replica found in json file, but no match for 'action' (%s)",
-                                           rse_key, replica_key, recoverable_replicas[vo][rse_key][replica_key]['surl'], i["datatype"], i["action"])
-                                break
-                            else:
-                                # If no policy has been set, default to ignoring the file (no action taken).
+                            # Check if a policy has been set for the specific datatype or scope
+                            # or if there are any wildcard policies regarding the scope.
+                            # Datatype action has priority over scope action.
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py:368]: I wonder if it should be this way. Let's rediscuss later

File: etc/suspicious_replica_recoverer.json [hunk 6]
```diff
@@ -1,10 +1,20 @@
...
-        "action": "declare bad"
+        "datatype_action": "ignore",
+        "datatype": ["RAW"]
+    },
+    {
+        "scope_action": "declare bad",
+        "scope": [],
+        "scope_wildcard": []
+    },
+    {
```
Comment (cserf) [inline etc/suspicious_replica_recoverer.json:8]: Isn't it possible to have combined policy for scope and datatype, e.g. : 
```json
    {
        "action": "declare bad",
        "scope": ["mc.*", "data12_8TeV"],
        "datatype": ["DAOD_PHYS.*", "DAOD_SUSY1"]
    },
```

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 7]
```diff
@@ -338,49 +334,101 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
...
+                            # Specific scope has priority over wildcard scope.
+                            if "datatype" in i:
+                                if file_metadata_datatype in i["datatype"]:
+                                    try:
+                                        action = i["datatype_action"]
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Why not simply action ?

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 8]
```diff
@@ -338,49 +334,101 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
...
+                                        in_file_metadata_datatype = True
+                                        break
+                                    except:
+                                        logger(logging.WARNING, "RSE: %s, replica name %s, surl %s: Datatype %s found in the policy library, but an error occured while retrieving the policy assigned to it. No action will be taken.",
+                                               rse_key, replica_key, recoverable_replicas[vo][rse_key][replica_key]['surl'], file_metadata_datatype)
+                                        break
+                            if ("scope" in i) and not in_file_metadata_datatype:
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: The action should be taken only if `in_file_metadata_datatype` if `datatype` is defined in the policy and `in_file_scope` if `scope` is defined in the policy. Therefore I would remove `and not in_file_metadata_datatype`

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 9]
```diff
@@ -338,49 +334,101 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
...
+                                # print("file_metadata_scope in i[scope]:", file_metadata_scope in i["scope"])
+                                if file_metadata_scope in i["scope"]:
+                                    try:
+                                        action = i["scope_action"]
+                                        in_file_scope = True
+                                    except:
+                                        logger(logging.WARNING, "RSE: %s, replica name %s, surl %s: Scope %s found in the policy library, but an error occured while retrieving the policy assigned to it. No action will be taken.",
+                                               rse_key, replica_key, recoverable_replicas[vo][rse_key][replica_key]['surl'], file_scope)
+                                        break
+                            if ("scope_wildcard" in i) and not (in_file_metadata_datatype or in_file_scope):
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Merge `scope_wildcard` with scope `scope

**Review by cserf — CHANGES_REQUESTED** [review]

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py (hunk 4 above)
Comment (ChristophAmes) [reply lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py:365]: The loop in line 132 closes in line 135. The main loop for the logic starts at line 149.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py (hunk 7 above)
Comment (ChristophAmes) [reply lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Do you mean why bother with the "try"?

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py (hunk 8 above)
Comment (ChristophAmes) [reply lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: The "not" clause is what makes sure that a datatype action has priority over a scope action. I'll remain the variable to make it a bit clearer.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py (hunk 9 above)
Comment (ChristophAmes) [reply lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: The idea is that if a policy is set for a scope wildcard, but a single scope affected by that wildcard should have a different policy, then the policy for that single scope will take priority over the wildcard policy for that one scope.

**Review by cserf — DISMISSED** [review]
LGTM. Still missing a `datatype_wildcard` but it can go into another patch.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 10]
```diff
@@ -197,7 +193,7 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
 
             # RSEs that aren't available shouldn't have suspicious replicas showing up. Skip to next RSE.
             if not rse['availability_read'] and ((len(suspicious_replicas_avail_elsewhere) > 0) or (len(suspicious_replicas_last_copy) > 0)):
-                logger(logging.WARNING, "%s is not available (availability_read: %s), yet it has suspicious replicas. Please investigate. \n", rse_expr, rse['availability_read'])
+                logger(logging.WARNING, "%s is not available (availability: %s), yet it has suspicious replicas. Please investigate. \n", rse_expr, rse['availability'])
```
Comment (bari12) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Why did you change this back to `availability`? This will be removed this release so we would have to change it back again then.

Comment (ChristophAmes) [reply lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: I'm not sure what happened there - it still says `availability_read` on my end. I've fixed it.

**Review by bari12 — APPROVED** [review]

**Comment by bari12** [comment]
I only merged this against `master` and thus the 1.31 release in march.
For the cherry-pick to the release branch, there is lots of conflicts which I won't fix manually. If this is needed during `1.30.*` still, please submit a second PR with this change against `release-1.30` branch.
